File: libs/document-processor-next/src/runtime/__tests__/emission-registry.test.ts
import { describe, expect, it } from 'vitest';
import { Blue } from '@blue-labs/language';

import { EmissionRegistry } from '../emission-registry.js';

const blue = new Blue();

function nodeFrom(json: unknown) {
  return blue.jsonValueToNode(json);
}

describe('EmissionRegistry', () => {
  it('creates and reuses scope runtime contexts', () => {
    const registry = new EmissionRegistry();

    const contextA = registry.scope('/root');
    const contextB = registry.scope('/root');

    expect(contextA).toBe(contextB);
    expect(registry.existingScope('/missing')).toBeUndefined();
  });

  it('records root emissions in order', () => {
    const registry = new EmissionRegistry();
    const first = nodeFrom({ id: '1' });
    const second = nodeFrom({ id: '2' });

    registry.recordRootEmission(first);
    registry.recordRootEmission(second);

    expect(registry.rootEmissions()).toEqual([first, second]);
  });

  it('tracks terminated scopes and allows clearing', () => {
    const registry = new EmissionRegistry();
    const scope = registry.scope('/child');
    scope.finalizeTermination('GRACEFUL', null);

    expect(registry.isScopeTerminated('/child')).toBe(true);
    registry.clearScope('/child');
    expect(registry.isScopeTerminated('/child')).toBe(false);
  });
});

File: references/java-sdk/src/test/java/blue/language/processor/EmissionRegistryTest.java
package blue.language.processor;

import blue.language.model.Node;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;

class EmissionRegistryTest {

    @Test
    void createsAndReusesScopeRuntimeContexts() {
        EmissionRegistry registry = new EmissionRegistry();

        ScopeRuntimeContext contextA = registry.scope("/root");
        ScopeRuntimeContext contextB = registry.scope("/root");

        assertSame(contextA, contextB);
        assertNull(registry.existingScope("/missing"));
    }

    @Test
    void scopeTreatsNullAndEmptyAsRoot() {
        EmissionRegistry registry = new EmissionRegistry();
        ScopeRuntimeContext root = registry.scope("/");
        assertNotNull(root);
        assertSame(root, registry.scope(null));
        assertSame(root, registry.scope(""));
    }

    @Test
    void scopeNormalizesNonPointerInputs() {
        EmissionRegistry registry = new EmissionRegistry();
        ScopeRuntimeContext normalized = registry.scope("root");
        assertSame(normalized, registry.scope("/root"));
        assertNull(registry.existingScope("/missing"));
        assertSame(normalized, registry.existingScope("root"));
        registry.clearScope("root");
        assertNull(registry.existingScope("/root"));
    }

    @Test
    void terminationLookupAndClearUseNormalizedScope() {
        EmissionRegistry registry = new EmissionRegistry();
        ScopeRuntimeContext root = registry.scope("/");
        root.finalizeTermination(ScopeRuntimeContext.TerminationKind.GRACEFUL, "done");

        assertTrue(registry.isScopeTerminated(null));
        assertTrue(registry.isScopeTerminated(""));

        registry.clearScope("");
        assertNull(registry.existingScope("/"));
        assertFalse(registry.isScopeTerminated("/"));
    }

    @Test
    void recordsRootEmissionsInOrder() {
        EmissionRegistry registry = new EmissionRegistry();
        Node first = new Node().properties("id", new Node().value("1"));
        Node second = new Node().properties("id", new Node().value("2"));

        registry.recordRootEmission(first);
        registry.recordRootEmission(second);

        assertEquals(Arrays.asList(first, second), registry.rootEmissions());
    }

    @Test
    void tracksTerminatedScopesAndAllowsClearing() {
        EmissionRegistry registry = new EmissionRegistry();
        ScopeRuntimeContext child = registry.scope("/child");
        child.finalizeTermination(ScopeRuntimeContext.TerminationKind.GRACEFUL, null);

        assertTrue(registry.isScopeTerminated("/child"));
        registry.clearScope("/child");
        assertFalse(registry.isScopeTerminated("/child"));
    }
}
